File: backend/src/CurriculosProIA.Service/Implementations/CoverLetterService.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using CurriculosProIA.Domain.Entities;
using CurriculosProIA.Domain.Dtos;
using CurriculosProIA.Repository.Persistence;
using CurriculosProIA.Domain.Dtos;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;

using CurriculosProIA.Service.Interfaces;
using CurriculosProIA.Repository.Interfaces;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace CurriculosProIA.Service.Implementations;

public class CoverLetterService : ICoverLetterService
{
    private readonly IAiService _aiService;
    private readonly IJobSitesService _jobSites;

    public CoverLetterService(IAiService aiService, IJobSitesService jobSites)
    {
        _aiService = aiService;
        _jobSites = jobSites;
        QuestPDF.Settings.License = LicenseType.Community;
    }

    public async Task<string> GenerateCoverLetterAsync(
        string resumeText,
        AnalysisInput analysis,
        string? siteId = null,
        CancellationToken cancellationToken = default)
    {
        var siteInfo = await BuildSiteInfoAsync(siteId, cancellationToken);
        var siteKeywords = new List<string>();

        if (!string.IsNullOrEmpty(siteId))
        {
            var site = await _jobSites.GetJobSiteByIdAsync(siteId, cancellationToken);
            siteKeywords = site?.PalavrasChavePadrao ?? new List<string>();
        }

        var habilidades = analysis.Habilidades != null ? string.Join(", ", analysis.Habilidades) : "Não especificado";
        var pontosFortes = analysis.PontosFortes != null ? string.Join(", ", analysis.PontosFortes) : "Não especificado";

        var systemPrompt = $"""
            Você é um especialista em redação de cartas de apresentação profissionais otimizadas para análise por sistemas ATS (Applicant Tracking Systems) e IAs de validação de currículo.
            Sua função é criar cartas de apresentação personalizadas, persuasivas e profissionais que destaquem as qualificações do candidato de forma estratégica e otimizada para passar por sistemas automatizados de triagem.

            IMPORTANTE:
            - A carta deve ser profissional, concisa e impactante
            - Destaque os pontos fortes identificados na análise
            - Use linguagem adequada ao contexto do site de vagas (se fornecido)
            - A carta deve ter entre 3-4 parágrafos
            - Seja específico e evite clichês genéricos
            - Mencione conquistas e resultados quando possível
            - Otimize a carta para passar por sistemas ATS e análise de IA de recrutadores
            {(siteKeywords.Count > 0 ? $"- Use naturalmente e estrategicamente as seguintes palavras-chave CRÍTICAS para o site: {string.Join(", ", siteKeywords)}" : "")}
            """;

        var userPrompt = $"""
            Com base no currículo e na análise fornecidos, crie uma carta de apresentação profissional e personalizada.

            {siteInfo}

            CURRÍCULO DO CANDIDATO:
            {resumeText}

            ANÁLISE DO CURRÍCULO:
            - Pontos Fortes: {pontosFortes}
            - Experiência: {analysis.Experiencia ?? "Não especificado"}
            - Formação: {analysis.Formacao ?? "Não especificado"}
            - Habilidades: {habilidades}
            - Score: {analysis.Score?.ToString() ?? "N/A"}/100

            Crie uma carta de apresentação que:
            1. Apresenta o candidato de forma profissional
            2. Destaca os principais pontos fortes e experiências relevantes
            3. Demonstra interesse e adequação para oportunidades
            4. Usa linguagem persuasiva mas profissional
            5. É específica e evita generalidades
            {(siteKeywords.Count > 0 ? $"6. Incorpora NATURALMENTE e ESTRATEGICAMENTE as palavras-chave CRÍTICAS: {string.Join(", ", siteKeywords)}" : "")}
            7. Está otimizada para passar por sistemas ATS e análise de IA de recrutadores

            Formato da carta:
            - Saudação profissional
            - Parágrafo introdutório: Apresentação e objetivo
            - Parágrafo(s) do meio: Destaque de qualificações e experiências relevantes
            - Parágrafo final: Encerramento profissional e disponibilidade para contato

            Retorne APENAS o texto da carta de apresentação, sem explicações adicionais.
            """;

        var letter = await _aiService.GenerateTextAsync($"{systemPrompt}\n\n{userPrompt}", 0.8, 1500, cancellationToken);
        return AiService.CleanMarkdownFence(letter);
    }

    public byte[] GenerateCoverLetterPdf(string coverLetterText)
    {
        var dateStr = DateTime.Now.ToString("dd 'de' MMMM 'de' yyyy", new CultureInfo("pt-BR"));
        var paragraphs = coverLetterText
            .Split(new[] { "\n\n", "\r\n\r\n" }, StringSplitOptions.RemoveEmptyEntries)
            .Select(p => p.Trim())
            .Where(p => !string.IsNullOrEmpty(p))
            .ToList();

        return Document.Create(container =>
        {
            container.Page(page =>
            {
                page.Size(PageSizes.A4);
                page.Margin(50);
                page.DefaultTextStyle(x => x.FontSize(11).FontFamily("Helvetica"));

                page.Content().Column(column =>
                {
                    column.Item().AlignCenter().Text("CARTA DE APRESENTAÇÃO").Bold().FontSize(18);
                    column.Item().PaddingVertical(15);
                    column.Item().AlignRight().Text(dateStr).FontSize(10);
                    column.Item().PaddingVertical(10);

                    for (var i = 0; i < paragraphs.Count; i++)
                    {
                        column.Item().Text(paragraphs[i]).Justify();
                        column.Item().PaddingBottom(10);
                    }

                    column.Item().PaddingTop(20).Text("Atenciosamente,");
                    column.Item().PaddingTop(15).Text("___________________________");
                });
            });
        }).GeneratePdf();
    }

    private async Task<string> BuildSiteInfoAsync(string? siteId, CancellationToken cancellationToken)
    {
        if (string.IsNullOrEmpty(siteId))
        {
            return string.Empty;
        }

        var site = await _jobSites.GetJobSiteByIdAsync(siteId, cancellationToken);
        if (site == null)
        {
            return string.Empty;
        }

        var keywords = site.PalavrasChavePadrao ?? new List<string>();
        var characteristics = site.Caracteristicas != null
            ? JsonSerializer.Serialize(site.Caracteristicas, new JsonSerializerOptions { WriteIndented = true })
            : "{}";

        var sb = new StringBuilder();
        sb.AppendLine("═══════════════════════════════════════════════════════════════");
        sb.AppendLine("CONTEXTO CRÍTICO - SITE DE VAGAS SELECIONADO:");
        sb.AppendLine($"Esta carta será usada no site: {site.Nome}");
        if (!string.IsNullOrEmpty(site.Descricao))
        {
            sb.AppendLine($"Descrição do site: {site.Descricao}");
        }

        sb.AppendLine($"Características específicas do site: {characteristics}");
        if (keywords.Count > 0)
        {
            sb.AppendLine($"PALAVRAS-CHAVE PRIORITÁRIAS PARA ESTE SITE (ESSENCIAIS PARA ATS): {string.Join(", ", keywords)}");
        }

        sb.AppendLine("═══════════════════════════════════════════════════════════════");
        return sb.ToString();
    }
}

File: backend-node/src/services/cover-letter.service.js
import '../load-env.js';
import OpenAI from 'openai';
import PDFDocument from 'pdfkit';
import { getJobSiteById } from './job-sites.service.js';

// OpenAI DESATIVADO temporariamente para não consumir créditos
// const openai = new OpenAI({
//   apiKey: process.env.OPENAI_API_KEY
// });
const openai = null; // Forçado para null para desativar OpenAI

const DEFAULT_MODEL = process.env.OPENAI_MODEL || 'gpt-4';

/**
 * Gera uma carta de apresentação personalizada baseada no currículo e no site de vagas
 */
export const generateCoverLetter = async (resumeText, analysis, siteId = null) => {
  try {
    let siteInfo = '';
    let siteKeywords = [];
    let siteCharacteristics = {};

    // Busca informações do site de vagas se fornecido
    if (siteId) {
      console.log(`📍 Buscando informações do site de vagas: ${siteId}`);
      try {
        const site = await getJobSiteById(siteId);
        if (site) {
          console.log(`✅ Site encontrado: ${site.nome}`);
          console.log(`🔑 Palavras-chave: ${Array.isArray(site.palavras_chave_padrao) ? site.palavras_chave_padrao.join(', ') : 'Nenhuma'}`);
          siteKeywords = Array.isArray(site.palavras_chave_padrao) ? site.palavras_chave_padrao : [];
          siteCharacteristics = site.caracteristicas && typeof site.caracteristicas === 'object' ? site.caracteristicas : {};
          
          siteInfo = `
═══════════════════════════════════════════════════════════════
CONTEXTO CRÍTICO - SITE DE VAGAS SELECIONADO:
Esta carta será usada no site: ${site.nome}
${site.descricao ? `Descrição do site: ${site.descricao}` : ''}
${site.caracteristicas ? `Características específicas do site: ${JSON.stringify(site.caracteristicas, null, 2)}` : ''}
${siteKeywords.length > 0 ? `PALAVRAS-CHAVE PRIORITÁRIAS PARA ESTE SITE (ESSENCIAIS PARA ATS): ${siteKeywords.join(', ')}` : ''}
═══════════════════════════════════════════════════════════════

IMPORTANTE: Toda a geração DEVE ser adaptada especificamente para o site ${site.nome}.
As palavras-chave acima são CRÍTICAS e devem ser incorporadas naturalmente no texto.
`;
        }
      } catch (error) {
        console.warn('⚠️ Erro ao buscar informações do site:', error.message);
      }
    }

    const systemPrompt = `Você é um especialista em redação de cartas de apresentação profissionais otimizadas para análise por sistemas ATS (Applicant Tracking Systems) e IAs de validação de currículo.
Sua função é criar cartas de apresentação personalizadas, persuasivas e profissionais que destaquem as qualificações do candidato de forma estratégica e otimizada para passar por sistemas automatizados de triagem.

IMPORTANTE:
- A carta deve ser profissional, concisa e impactante
- Destaque os pontos fortes identificados na análise
- Use linguagem adequada ao contexto do site de vagas (se fornecido)
- A carta deve ter entre 3-4 parágrafos
- Seja específico e evite clichês genéricos
- Mencione conquistas e resultados quando possível
- Adapte o tom e estilo conforme o site de vagas selecionado
- Otimize a carta para passar por sistemas ATS e análise de IA de recrutadores
${siteKeywords.length > 0 ? `- Use naturalmente e estrategicamente as seguintes palavras-chave CRÍTICAS para o site: ${siteKeywords.join(', ')} - estas palavras são essenciais para passar por sistemas ATS` : ''}
${siteId ? `- A carta DEVE ser especificamente adaptada para o site de vagas selecionado` : ''}`;

    const userPrompt = `Com base no currículo e na análise fornecidos, crie uma carta de apresentação profissional e personalizada.

${siteInfo}

CURRÍCULO DO CANDIDATO:
${resumeText}

ANÁLISE DO CURRÍCULO:
- Pontos Fortes: ${Array.isArray(analysis.pontosFortes) ? analysis.pontosFortes.join(', ') : (analysis.pontosFortes || 'Não especificado')}
- Experiência: ${analysis.experiencia || 'Não especificado'}
- Formação: ${analysis.formacao || 'Não especificado'}
- Habilidades: ${Array.isArray(analysis.habilidades) ? analysis.habilidades.join(', ') : (analysis.habilidades || 'Não especificado')}
- Score: ${analysis.score || 'N/A'}/100

${siteId ? `IMPORTANTE: Esta carta será usada no site ${siteInfo.includes('Nome:') ? siteInfo.split('Nome:')[1].split('\n')[0].trim() : 'selecionado'}. Adapte o conteúdo, tom e palavras-chave para este contexto específico.` : ''}

Crie uma carta de apresentação que:
1. Apresenta o candidato de forma profissional
2. Destaca os principais pontos fortes e experiências relevantes
3. Demonstra interesse e adequação para oportunidades
4. Usa linguagem persuasiva mas profissional
5. É específica e evita generalidades
${siteKeywords.length > 0 ? `6. Incorpora NATURALMENTE e ESTRATEGICAMENTE as palavras-chave CRÍTICAS: ${siteKeywords.join(', ')} - estas são ESSENCIAIS para passar por sistemas ATS e análise de IA` : ''}
7. Está otimizada para passar por sistemas ATS e análise de IA de recrutadores
${siteId ? `8. É especificamente adaptada para o site ${siteInfo.includes('Nome:') ? siteInfo.split('Nome:')[1].split('\n')[0].trim() : 'selecionado'} - use o contexto e características deste site` : ''}

Formato da carta:
- Saudação profissional (Ex: "Prezados Senhores," ou "Caro(a) Recrutador(a),")
- Parágrafo introdutório: Apresentação e objetivo
- Parágrafo(s) do meio: Destaque de qualificações e experiências relevantes
- Parágrafo final: Encerramento profissional e disponibilidade para contato

Retorne APENAS o texto da carta de apresentação, sem explicações adicionais.`;

    console.log('🤖 Gerando carta de apresentação com IA...');

    if (!openai) {
      throw new Error('OpenAI está desativado temporariamente. Use Gemini ou ative o modo mock.');
    }

    const completion = await openai.chat.completions.create({
      model: DEFAULT_MODEL,
      messages: [
        {
          role: "system",
          content: systemPrompt
        },
        {
          role: "user",
          content: userPrompt
        }
      ],
      temperature: 0.8,
      max_tokens: 1500
    });

    const coverLetter = completion.choices[0].message.content.trim();
    
    // Remove markdown se existir
    let cleanedLetter = coverLetter;
    if (cleanedLetter.startsWith('```')) {
      cleanedLetter = cleanedLetter.replace(/^```[a-z]*\s*/, '').replace(/\s*```$/, '');
    }

    console.log('✅ Carta de apresentação gerada com sucesso');
    return cleanedLetter;

  } catch (error) {
    console.error('❌ Erro ao gerar carta de apresentação:', error);
    throw new Error(`Erro ao gerar carta de apresentação: ${error.message}`);
  }
};

/**
 * Converte texto da carta de apresentação em PDF
 */
export const generateCoverLetterPDF = (coverLetterText) => {
  return new Promise((resolve, reject) => {
    try {
      const doc = new PDFDocument({
        size: 'A4',
        margins: { top: 50, bottom: 50, left: 50, right: 50 }
      });

      const chunks = [];
      doc.on('data', chunk => chunks.push(chunk));
      doc.on('end', () => resolve(Buffer.concat(chunks)));
      doc.on('error', reject);

      // Título
      doc.fontSize(18).font('Helvetica-Bold').text('CARTA DE APRESENTAÇÃO', { align: 'center' });
      doc.moveDown(2);

      // Data (opcional, pode ser adicionada dinamicamente)
      const today = new Date();
      const dateStr = today.toLocaleDateString('pt-BR', { 
        day: '2-digit', 
        month: 'long', 
        year: 'numeric' 
      });
      doc.fontSize(10).font('Helvetica').text(dateStr, { align: 'right' });
      doc.moveDown(1.5);

      // Processa o texto da carta
      const paragraphs = coverLetterText.split(/\n\s*\n/).filter(p => p.trim());
      
      paragraphs.forEach((paragraph, index) => {
        const trimmedParagraph = paragraph.trim();
        
        if (!trimmedParagraph) {
          doc.moveDown(0.5);
          return;
        }

        // Primeiro parágrafo pode ter saudação
        if (index === 0 && trimmedParagraph.length < 100) {
          doc.fontSize(11).font('Helvetica').text(trimmedParagraph);
          doc.moveDown(1);
        } else {
          // Parágrafos normais
          doc.fontSize(11).font('Helvetica').text(trimmedParagraph, {
            align: 'justify',
            width: 500,
            lineGap: 3
          });
          doc.moveDown(1);
        }
      });

      // Espaço para assinatura
      doc.moveDown(2);
      doc.fontSize(11).font('Helvetica').text('Atenciosamente,', { align: 'left' });
      doc.moveDown(1.5);
      doc.fontSize(11).font('Helvetica').text('___________________________', { align: 'left' });

      doc.end();
    } catch (error) {
      reject(new Error(`Erro ao gerar PDF da carta: ${error.message}`));
    }
  });
};
